adam-osu/job-tracker
# Add db

## What this PR does

- Adds docker-compose to to run postgres and adminer
- Sets up a connection to the database running as a container
- Scaffolds the user model via a [migration](https://sequelize.org/docs/v6/other-topics/migrations/) and adds a beforeCreate hook to hash and salt the password
- Seeds a single user using sequelize-cli
- Updates readme

## Getting started

### Docker

1. Download and install Docker desktop
2. From the command line run `docker volume create pgdata`
3. Run `docker-compose up --build` to run Postgres and Adminer at http://localhost:8080
   If you get this error:

  ```
  PostgreSQL Database directory appears to contain a database; Skipping initialization
  ```

   Run `docker-compose down --volumes` and try `docker-compose up --build` again
4. Log in
<img width="265" alt="pglogin" src="https://user-images.githubusercontent.com/62804095/162624161-01abece1-906d-4b87-bcf7-a3d6dc7157da.png">

### Tophatting (Testing) locally

1. Make sure step above (running postgres in docker container) was successful 
2. Run the pending migration (to create Users table) with `npx sequelize-cli db:migrate`. 

<img width="530" alt="Screen Shot 2022-04-10 at 9 40 41 AM" src="https://user-images.githubusercontent.com/62804095/162626104-8c21c7ff-630b-4296-9f6d-1e3489264936.png">

3. Visit adminer at localhost:8080 and look for the newly created table

<img width="493" alt="Screen Shot 2022-04-10 at 10 35 27 AM" src="https://user-images.githubusercontent.com/62804095/162624538-cfa4e423-444b-4940-9a1d-abbfdee6e91f.png">

4. Start up the server in development using `npm run start:dev`
5. Run the seed command to create a fake user with `npx sequelize-cli db:seed:all`
6. Check the db once again using adminer to look for the newly created column

<img width="1105" alt="Screen Shot 2022-04-10 at 10 12 07 AM" src="https://user-images.githubusercontent.com/62804095/162624889-2787ed91-14d5-4c22-98b6-969d5e392d3d.png">


### About migrations

To create and update our db tables and bootstrap our models, we will be using migrations. The sequelize-cli tool included in our project helps to generate the migration files and run them and in additional bootstraps our models.

1. Please read the introductory section [here](https://sequelize.org/docs/v6/other-topics/migrations)
6. How [create migrations](https://sequelize.org/docs/v6/other-topics/migrations/#creating-the-first-model-and-migration) 
8. How to [run migrations](https://sequelize.org/docs/v6/other-topics/migrations/#running-migrations)
9. How to [undo migrations](https://sequelize.org/docs/v6/other-topics/migrations/#undoing-migrations)


In this PR, I've created a migration file for Users and ran it to update the table using:
```
npx sequelize-cli model:generate --name User --attributes name:string,email:string,password:string
```
(As a note I made some modifications to the model file directly, which you may sometimes need to do).

The command to `npx sequelize-cli db:migrate` also bootstrapped the [User model](https://github.com/adam-osu/job-tracker/blob/62771422459bdfa1c29d6b79894a929cff07cb45/models/user.js)!

<img width="530" alt="Screen Shot 2022-04-10 at 9 40 41 AM" src="https://user-images.githubusercontent.com/62804095/162624555-60705966-102a-465e-850b-6f0c81bf00ac.png">

When I log in to adminer, I can now see the table was created.

<img width="493" alt="Screen Shot 2022-04-10 at 10 35 27 AM" src="https://user-images.githubusercontent.com/62804095/162624538-cfa4e423-444b-4940-9a1d-abbfdee6e91f.png">


### About seeds

You can also create seed (dummy) data using sequelize-cli

1. How to [create seeds](https://sequelize.org/docs/v6/other-topics/migrations/#creating-the-first-model-and-migration)
3. How to [run seeds](https://sequelize.org/docs/v6/other-topics/migrations/#running-seeds) 
4. How to [undo seeds](https://sequelize.org/docs/v6/other-topics/migrations/#undoing-seeds)

For users I have generated a seed file and ran it to create a column in the Users table:
<img width="1105" alt="Screen Shot 2022-04-10 at 10 12 07 AM" src="https://user-images.githubusercontent.com/62804095/162624889-2787ed91-14d5-4c22-98b6-969d5e392d3d.png">
